### PR TITLE
Fix and enhance `Mobject.arrange_in_grid`

### PR DESCRIPTION
**Pull request by AntonBallmaier**
## Changelog / Overview

`arrange_in_grid` now actually arranges submobjects in a grid. Added new parameters `buff`, `cell_alignment`, `row_alignments`, `col_alignments`, `row_heights`, `col_widths`, `flow_order`.

## Motivation
`arrange_in_grid` sounds like a useful method, until you use it and notice it doesn't arrange in a grid but simply in multiple independent rows. Therefor this is not working for differently sized mobjects. I fixed that and added all the options to grid alignment that I could think of :slightly_smiling_face: 

## Explanation for Changes
Have a look at the [docs](https://manimce--1398.org.readthedocs.build/en/1398/reference/manim.mobject.mobject.Mobject.html#manim.mobject.mobject.Mobject.arrange_in_grid) :)

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by AntonBallmaier** [comment]
Is there a better way to restart the CI than just to reopen the PR? The docs building simply fails when ever it wants...

**Comment by abhi3019** [comment]
@AntonBallmaier You can create an empty commit and push
```py
git commit --allow-empty -m "Empty commit or some other commit message as you wish"
git push
```
Source : https://stackoverflow.com/questions/20138640/pushing-empty-commits-to-remote

**Comment by hydrobeam** [comment]
I've had quite a bit of annoyance trying to use `arrange_in_grid` before so this is nice!

At first glance these changes seem rather overwhelming, but I they are rather comprehensive. In my opinion, having both `cell_alignment` and `row/col_alignment` is a bit redundant since I don't think that level is control is needed for most use cases of `arrange_in_grid`, although it might be me. Plus, using `l/c/r` does seem slightly odd in terms of syntax. An alternative might be to have a way to control the orientation of each cell using an array?

I'm quite a fan of `flow_order`! But, (not a deal breaker by any means) it would be nice to introduce a snake-y method of adding objects?:  (a a 7 where the 12 is)
image

File: manim/mobject/mobject.py [hunk 1]
```diff
@@ -2039,40 +2040,236 @@ def construct(self):
...
+            columns rigthwards. "u" and "l" stand for upwards and leftwards.
+
+        .. note:: If either :code:`rows` or :code:`cols` is not defined, it is set to the length of either
+            :code:`row_alignments` / :code:`col_alignments` or :code:`row_heights` / :code:`col_widths` if
+            one of those is defined. If only one of :code:`rows` and :code:`cols` can be determined
+            like this, the other will be choosen big enough to fit all submobjects.
+            If neither :code:`cols` nor :code:`rows` is set implicitly, they will be choosen to be
+            about the same, tending towards :code:`cols > rows` (simply because videos are wider
+            then they are high).
+
```
Comment (hydrobeam) [inline manim/mobject/mobject.py:2100]: In my opinion this information is a little irrelevant for most users and can be determined through some light experimentation. It definitely feels more dev-facing. Maybe it would work better as an in-code comment?
```suggestion
```

File: manim/mobject/mobject.py [hunk 2]
```diff
@@ -2039,40 +2040,236 @@ def construct(self):
...
+        .. note:: If both :code:`cell_alignment` and either :code:`row_alignments` or :code:`col_alignments` is
+            defined the latter has higher priority in its direction.
```
Comment (hydrobeam) [inline manim/mobject/mobject.py:2102]:
```suggestion
        .. note:: If both :code:`cell_alignment` and :code:`row_alignments` / :code:`col_alignments` are
            defined, the latter has higher priority.
```

File: manim/mobject/mobject.py [hunk 3]
```diff
@@ -2039,40 +2040,236 @@ def construct(self):
...
+
+        Returns
+        -------
+        Mobject
+            The mobject.
+
+        Raises
+        ------
+        ValueError
+            If rows and cols are choosen to small to fit all submobjetcs.
```
Comment (hydrobeam) [inline manim/mobject/mobject.py:2113]:
```suggestion
            If :attr:`rows` and :attr:`cols` are too small to fit all submobjects.
```

File: manim/mobject/mobject.py [hunk 4]
```diff
@@ -2039,40 +2040,236 @@ def construct(self):
...
+        ValueError
+            :code:`cols`, :code:`col_alignments` and :code:`col_widths` or :code:`rows`,
+            :code:`row_alignments` and :code:`row_heights` have mismatching sizes.
```
Comment (hydrobeam) [inline manim/mobject/mobject.py:2116]: Just to be consistent with the previous ValueError
```suggestion
        ValueError
            If :code:`cols`, :code:`col_alignments` and :code:`col_widths` or :code:`rows`,
            :code:`row_alignments` and :code:`row_heights` have mismatching sizes.
```

**Review by hydrobeam — CHANGES_REQUESTED** [review]
I've left some comments/suggestions mainly about the documentation to make it easier for a user to understand how the method works and not feel overwhelmed: less text overall + slight grammar fixes.

File: manim/mobject/mobject.py [hunk 5]
```diff
@@ -2039,40 +2040,236 @@ def construct(self):
...
+                        buff=(0.25,0.5),
+                        col_alignments="lccccr",
+                        row_alignments="uccd",
+                        col_widths=[2, *[None]*4, 2],
+                        flow_order="dr"
+                    )
+
+
         """
+        from manim import Line
```
Comment (kilacoda) [inline manim/mobject/mobject.py:2141]: AFAIK, the default policy for intra-library imports is to use relative imports to the required module, like the suggestion.
```suggestion
        from .geometry import Line
```

File: manim/mobject/mobject.py [hunk 6]
```diff
@@ -2039,40 +2040,236 @@ def construct(self):
...
+            on the highest element in that row.
+        col_widths
+            Specific widths for certain columns (from left to right). The given list can contain
+            `None`. In that case the corresponding column will fit its width automatically based
+            on the widest element in that column.
+        flow_order
+            The order in which the grid gets filled with the submobjects. Can be one of the
+            following values: "rd", "dr", "ld", "dl", "ru", "ur", "lu", "ul". For example,
+            "dr" (down right) means: Fill the table column by column downwards, filling the
+            columns rigthwards. "u" and "l" stand for upwards and leftwards.
```
[the remark this answers is not in the document]
Comment (AntonBallmaier) [reply manim/mobject/mobject.py:2091]: I feel like your suggestion is lacking the information what the difference between "rd" and "dr" is... I think the extra text is required for a precise explanation.

File: manim/mobject/mobject.py (hunk 1 above)
Comment (AntonBallmaier) [reply manim/mobject/mobject.py:2100]: Without it a user might not get the idea to leave cols or rows undefined, therefor I think the note is an improvement.

**Comment by AntonBallmaier** [comment]
> In my opinion, having both `cell_alignment` and `row/col_alignment` is a bit redundant since I don't think that level is control is needed for most use cases of `arrange_in_grid`, although it might be me.

Yes, it is redundant, but in my opinion really useful: Using `row/col_alignment` tables can easily be structured, `cell_alignment` is useful in less specific usecases. 

> An alternative might be to have a way to control the orientation of each cell using an array?

I would agree to add this only if that doesn't imply removing both `cell_alignment` or `row/col_alignment`. Imagine wanting to make a table, where you want to set the general alignment or just row- or column-wise and having to generate a full matrix... Would be rather tedious.

> Plus, using `l/c/r` does seem slightly odd in terms of syntax. 

That is the syntax latex uses for tables.

**Comment by AntonBallmaier** [comment]
> I'm quite a fan of `flow_order`! But, (not a deal breaker by any means) it would be nice to introduce a snake-y method of adding objects?: (a a 7 where the 12 is)

I am unsure about this. If we were to add this, where do we draw the line? In some other cases it might be useful to add the mobjects spiraling inwards, in other still to be ordered randomly and so on. I can think of two reasonable ways to solve this problem:
* Add an option to provide a function defining the row and column for each index. This covers all the above usecases.
* Just leave it as it is. The order of mobjects can  still be changed by the user afterwards without moving them around.

File: manim/mobject/mobject.py (hunk 1 above)
Comment (hydrobeam) [reply manim/mobject/mobject.py]: Maybe a middle ground would be to drop the second half of the note to keep the note concise and informative?: 
> If neither :code:`cols` nor :code:`rows` is set implicitly, they will be choosen to be
            about the same, tending towards :code:`cols > rows` (simply because videos are wider
            then they are high).

File: manim/mobject/mobject.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hydrobeam) [reply manim/mobject/mobject.py:2085]: I feel like it's rather intuitive how to interpret it, but I see what you mean. Maybe rewording to `"rd" --> fill rightwards then downwards` would me a more concise way of phrasing the explanation?

File: manim/mobject/mobject.py [hunk 7]
```diff
@@ -2039,40 +2040,235 @@ def construct(self):
...
+            The gap between grid cells. To specify a different buffer in the horizontal and
+            vertical directions, a tuple of two values can be given - `(row, col)`.
+        cell_alignment
+            The way each submobject is aligned in its grid cell.
+        row_alignments
+            The vertical alignment for each row (top to bottom). Accepts the following characters: `"u"` -
+            up, `"c"` - center, `"d"` - down.
+        col_alignments
+            The horizontal alignment for each column (left to right). Accepts the following characters `"l"` - left,
+             `"c"` - center, `"r"` - right.
```
Comment (hydrobeam) [inline manim/mobject/mobject.py:2074]: In the rendered page, there is some unfortunate formatting going on with this (I think I might have introduced the error 😅). This should fix it.
```suggestion
        col_alignments
            The horizontal alignment for each column (left to right). Accepts the following characters `"l"` - 
            left, `"c"` - center, `"r"` - right.
```

**Comment by hydrobeam** [comment]
> I am unsure about this. If we were to add this, where do we draw the line? In some other cases it might be useful to add the mobjects spiraling inwards, in other still to be ordered randomly and so on. I can think of two reasonable ways to solve this problem:

This is a fair concern. It *would* be a little overboard to cover every use case.

> - Just leave it as it is. The order of mobjects can still be changed by the user afterwards without moving them around.

Is there an easy way to edit afterwads? It seems like the modifications would have to happen to the submobjects of the `VGroup/Group` before the arrangment. I do agree with leaving it as is to avoid unnecessary complexity.

File: manim/mobject/mobject.py (hunk 1 above)
Comment (AntonBallmaier) [reply manim/mobject/mobject.py]: How about:
> If only one of :code:cols and :code:rows is set implicitly, the other one will be chosen big enough to fit all submobjects. If neither is set, they will be chosen to be about the same, tending towards :code:cols > rows (simply because videos are wider than they are high).

**Comment by AntonBallmaier** [comment]
> Is there an easy way to edit afterwads? It seems like the modifications would have to happen to the submobjects of the `VGroup/Group` before the arrangment.

No, you would have to reorder the submobject list after arrangement.

File: manim/mobject/mobject.py (hunk 1 above)
Comment (hydrobeam) [reply manim/mobject/mobject.py]: That seems acceptable

**Comment by hydrobeam** [comment]
The parameters section seems to duplicated? Once with docstrings and once without. The dupe appears at the end after the Notes, but I'm not sure what's causing that effect.

**Comment by AntonBallmaier** [comment]
> The parameters section seems to duplicated? Once with docstrings and once without. The dupe appears at the end after the Notes and before the example, but I'm not sure what's causing that effect.

You are right. Really wired.

File: manim/mobject/mobject.py [hunk 8]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
+
+        Parameters
+        ----------
+        rows
+            The number of rows in the grid.
+        cols
+            The number of columns in the grid.
+        buff
+            The gap between grid cells. To specify a different buffer in the horizontal and
+            vertical directions, a tuple of two values can be given - `(row, col)`.
```
Comment (behackl) [inline manim/mobject/mobject.py:2067]:
```suggestion
            vertical directions, a tuple of two values can be given - ``(row, col)``.
```

File: manim/mobject/mobject.py [hunk 9]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
+        cell_alignment
+            The way each submobject is aligned in its grid cell.
+        row_alignments
+            The vertical alignment for each row (top to bottom). Accepts the following characters: `"u"` -
+            up, `"c"` - center, `"d"` - down.
```
Comment (behackl) [inline manim/mobject/mobject.py:2072]:
```suggestion
            The vertical alignment for each row (top to bottom). Accepts the following characters: ``"u"`` -
            up, ``"c"`` - center, ``"d"`` - down.
```

File: manim/mobject/mobject.py [hunk 10]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
+        col_alignments
+            The horizontal alignment for each column (left to right). Accepts the following characters `"l"` - left,
+            `"c"` - center, `"r"` - right.
```
Comment (behackl) [inline manim/mobject/mobject.py:2075]:
```suggestion
            The horizontal alignment for each column (left to right). Accepts the following characters ``"l"`` - left,
            ``"c"`` - center, ``"r"`` - right.
```

File: manim/mobject/mobject.py [hunk 11]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
+        row_heights
+            Defines a list of heights for certain rows (top to bottom). If the list contains `None`, the corresponding row will fit its height automatically based
```
Comment (behackl) [inline manim/mobject/mobject.py:2077]:
```suggestion
            Defines a list of heights for certain rows (top to bottom). If the list contains 
            ``None``, the corresponding row will fit its height automatically based
```

File: manim/mobject/mobject.py [hunk 12]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
+            on the highest element in that row.
+        col_widths
+            Defines a list of widths for certain columns (left to right). If the list contains `None`, the
```
Comment (behackl) [inline manim/mobject/mobject.py:2080]:
```suggestion
            Defines a list of widths for certain columns (left to right). If the list contains ``None``, the
```

File: manim/mobject/mobject.py [hunk 13]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
+        -----
+
+        .. note::
+            If only one of :code:cols and :code:rows is set implicitly, the other one will be chosen big
+            enough to fit all submobjects. If neither is set, they will be chosen to be about the same,
+            tending towards :code:cols > rows (simply because videos are wider than they are high).
+
+        .. note::
+            If both :code:`cell_alignment` and :code:`row_alignments` / :code:`col_alignments` are
+            defined, the latter has higher priority.
```
Comment (behackl) [inline manim/mobject/mobject.py:2101]: note-blocks within the note-section are overkill.

```suggestion
        If only one of ``cols`` and ``rows`` is set implicitly, the other one will be chosen big
        enough to fit all submobjects. If neither is set, they will be chosen to be about the same,
        tending towards ``cols`` > ``rows`` (simply because videos are wider than they are high).

        If both ``cell_alignment`` and ``row_alignments`` / ``col_alignments`` are
        defined, the latter has higher priority.
```

File: manim/mobject/mobject.py [hunk 14]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
+
+
+        Raises
+        ------
+        ValueError
+            If :attr:`rows` and :attr:`cols` are too small to fit all submobjects.
```
Comment (behackl) [inline manim/mobject/mobject.py:2107]: Doesn't :attr: actually refer to attributes (`self.cols`) instead of keyword arguments? It should probably be changed to reflect that.

```suggestion
            If ``rows`` and ``cols`` are too small to fit all submobjects.
```

File: manim/mobject/mobject.py [hunk 15]
```diff
@@ -2040,40 +2041,229 @@ def construct(self):
...
 
-            class ArrangeExample(Scene):
+            class ArrangeInGrid(Scene):
                 def construct(self):
-                    self.add(*[Square(color= random_bright_color()) for i in range(0,5)])
-                    x = VGroup(*self.mobjects).arrange_in_grid(buff=0.2)
-                    self.add(x)
+                    #Add some numbered boxes:
+                    np.random.seed(3)
+                    boxes = VGroup(*[Rectangle(WHITE, np.random.random()+.5, np.random.random()+.5).add(Text(str(i+1)).scale(0.5)) for i in range(22)])
```
Comment (behackl) [inline manim/mobject/mobject.py:2122]: This line is a bit too long, please break.

**Review by behackl — CHANGES_REQUESTED** [review]
From reading the code, this seems fine to me -- thank you for your efforts! I've added some suggestions regarding the documentation: please remember that code literals in ReST are marked by double backticks.

**Review by behackl — APPROVED** [review]
